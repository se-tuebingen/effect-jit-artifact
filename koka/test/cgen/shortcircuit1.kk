effect choose
  ctl flip() : bool

fun mystery() : <choose, console> bool 
  val b = flip()
  println("b = " ++ b.show)
  b

// is `mystery` ever true?
fun satisfiable-no-short-circuit() : <console> bool 
  with ctl flip()
    // for each input flip(), try both values
    ( (resume(True) || resume(False)) : bool )
  mystery()

fun satisfiable-short-circuit() : <console> bool 
  with ctl flip()
    (resume(True) : bool) || resume(False)
  mystery()

fun main() 
  satisfiable-no-short-circuit().println
  satisfiable-short-circuit().println
