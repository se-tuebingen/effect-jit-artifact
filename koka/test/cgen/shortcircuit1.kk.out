b = True
True
b = True
True